Fix broken numeric text entry in state-card-input_number

karwosts opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change

The text field for number entry in "state-card-input_number" (and state-card-number) is pretty badly broken. 
1. You generally cannot enter 0's after a decimal because they are immediately erased, e.g. entering "3.03":

![303](https://user-images.githubusercontent.com/32912880/208248382-a8067d61-04b3-4d71-95d8-4a24c023a647.gif)

2. some rounding occurs as you type, so entering "0.555555555" ends up with something like "0.8". 

![number2](https://user-images.githubusercontent.com/32912880/208248411-cfccac2d-e259-4ba7-842a-f737b325a53c.gif)


I find that both of these problems stem from the hidden `<ha-slider>` that is instanced in this page. Currently input number entry can be selected to either be slider or textbox. The polymer template instances both, but hides the one that is not used. 

The problem is that the input slider, while hidden, is still watching the value change and so every time a number is entered it tries to snap the entry to match its slider step points. 

To resolve this, I wrap the slider in a `<template is="dom-if">` to remove the slider from the DOM when it is not used, not just hide it. This allows number entry to work as expected (textfield value exactly reflects keystrokes entered). 


## Type of change

<!--
  What type of change does your PR introduce to the Home Assistant frontend?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR.
-->

```yaml

```

## Additional information

<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #12220 
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

<!--
  Thank you for contributing <3
-->

[docs-repository]: https://github.com/home-assistant/home-assistant.io

karwosts committed:
Fix broken numeric text entry in state-card-input_number and state-card-number

bramkragten commented on src/state-summary/state-card-input_number.js:
```diff
@@ -42,19 +43,21 @@ class StateCardInputNumber extends mixinBehaviors(
-          ignore-bar-touch=""
-        >
-        </ha-slider>
+        <template is="dom-if" if="[[hiddenbox]]">
```
```suggestion
        <template is="dom-if" if="[[showslider]]">
```

bramkragten replied:
Let's do the same for the textfield

bramkragten commented on src/state-summary/state-card-input_number.js:
```diff
@@ -42,19 +43,21 @@ class StateCardInputNumber extends mixinBehaviors(
+            max="[[max]]"
+            value="{{value}}"
+            step="[[step]]"
+            hidden="[[hiddenslider]]"
```
```suggestion
```

karwosts commented:
Found this needs some work in the hiddenState() function. Wrapping the slider in a conditional means the width is not calculated the first time hiddenState is called, which is throwing an exception. 

Need to figure out a way to mitigate.

bramkragten commented:
Migrating this to lit will probably solve the issue, not sure that is something you want to do?

karwosts commented:
> Migrating this to lit will probably solve the issue, not sure that is something you want to do?

I came to the same conclusion, working on that now.

karwosts committed:
Fix bug in input_number entry, update from polymer to lit

karwosts commented:
Ok I switched the file to lit. This was my first time doing so, so there may be dumb mistakes in there. I borrowed a lot of the code from hui-input-number-entity-row which already did a lot of what I wanted. 

I also reverted the initial change to state-card-number. It probably could benefit from the same fixes, but I don't know any integrations that provide a number to test with that I could hookup to my dev environment. wled provides a number, but only sliders. 

So I will refrain from modifying that, since I can't test it. 

Final render:

![image](https://user-images.githubusercontent.com/32912880/209196359-a74d30f9-3ebf-4227-8f1b-e46ebc22e99f.png)
